### PR TITLE
cuda: exclude compute_30 when CUDA 11 is used

### PR DESCRIPTION
**Pull request by shintaro-iwasaki**
## Pull Request Description

CUDA 11 dropped `compute_30` support http://arnon.dk/matching-sm-architectures-arch-and-gencode-for-various-nvidia-cards/), but by default (without explicitly setting `--with-cuda-sm`), Yaksa tries to compile `compute_30`, which causes a compilation error. This patch fixes it by excluding `compute_30` when CUDA 11 is used.

For example, `compute_30` is not in allowed values.
```sh
$ nvcc --help
...
--gpu-architecture <arch>                       (-arch)
        ...
        Note: the values compute_30, compute_32, compute_35, compute_37, compute_50,
        sm_30, sm_32, sm_35, sm_37 and sm_50 are deprecated and may be removed in
        a future release.
        Allowed values for this option:  'compute_35','compute_37','compute_50',
        'compute_52','compute_53','compute_60','compute_61','compute_62','compute_70',
        'compute_72','compute_75','compute_80','lto_35','lto_37','lto_50','lto_52',
        'lto_53','lto_60','lto_61','lto_62','lto_70','lto_72','lto_75','lto_80',
        'sm_35','sm_37','sm_50','sm_52','sm_53','sm_60','sm_61','sm_62','sm_70',
        'sm_72','sm_75','sm_80'.
...
```

I have tested it on Summit with CUDA 11.0.3

## Expected Impact

None.

By default CUDA 11 shows many warnings while compiling Yaksa. Maybe we would like to exclude them when CUDA 11 is used.
```sh
nvcc warning : The 'compute_35', 'compute_37', 'compute_50', 'sm_35', 'sm_37' and 'sm_50' architectures are deprecated, and may be removed in a future release (Use -Wno-deprecated-gpu-targets to suppress warning).
```

## Author Checklist
* [x] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [x] Commits are self-contained and do not do two things at once
* [x] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [x] Add comments such that someone without knowledge of the code could understand
* [x] Have read and agree to the Yaksa CLA terms (https://github.com/pmodels/yaksa/wiki/Yaksa-Contributor-License-Agreement)

### CODE REVIEW

**Comment by pavanbalaji** [comment]
I think we can simply drop `compute_30` support in the autodetection case.  If users want, they can explicitly enable it anyway.

**Comment by shintaro-iwasaki** [comment]
Thanks. Although it is not a compilation error, but the following warning is also annoying if we compile Yaksa with CUDA 11.
```
nvcc warning : The 'compute_35', 'compute_37', 'compute_50', 'sm_35', 'sm_37' and 'sm_50'
```
It shows up so many times while compiling MPICH + CUDA support with CUDA11. Should we drop them as well as `compute_30` in the autodetection case for all CUDA versions? There are four options:
1. Drop `compute_30` in the case of the autodetection + all CUDA versions.
2. Drop `compute_30` in the case of the autodetection + CUDA 11+.
3. Drop `compute_30, 35, 37, 50` in the case of the autodetection + all CUDA versions.
4. Drop `compute_30, 35, 37, 50` in the case of the autodetection + CUDA 11+.

**Comment by pavanbalaji** [comment]
It looks like all of kepler support has been dropped in CUDA 11.  https://arnon.dk/matching-sm-architectures-arch-and-gencode-for-various-nvidia-cards/

I think we should change our detection logic based on the range of supported architectures in a given version, rather than just the minimum CUDA version.

**Comment by shintaro-iwasaki** [comment]
I updated the script. When CUDA11+ is used, Yaksa does not compile 30, 35, 37, and 50 by default since they are unsupported or deprecated. The detection part was also rewritten; I explicitly wrote supported SMs for each CUDA version

**Review by pavanbalaji — APPROVED** [review]
Looks great.  Thank you!

**Comment by pavanbalaji** [comment]
You might need to rebase it on to `main` before merging.

**Comment by shintaro-iwasaki** [comment]
Rebased.

**Comment by pavanbalaji** [comment]
@shintaro-iwasaki   It looks like you never merged this PR in.  Are you still waiting for something?  Or is it ready to be merged?

**Comment by shintaro-iwasaki** [comment]
It's ready. Can I merge it? If so I will tonight.

**Comment by pavanbalaji** [comment]
Can you rebase it on the main first?

**Comment by shintaro-iwasaki** [comment]
Rebased. I will merge.

**Comment by jeffhammond** [comment]
I am seeing a problem with `compute_30`.  I thought this was fixed already.

```
~/MPI/mpich/configure FCFLAGS=-fallow-argument-mismatch
```

```
Making all in modules/yaksa
make[2]: Entering directory '/tmp/modules/yaksa'
  NVCC     src/backend/cuda/pup/yaksuri_cudai_pup__Bool.lo
nvcc fatal   : Unsupported gpu architecture 'compute_30'
make[2]: *** [Makefile:14154: src/backend/cuda/pup/yaksuri_cudai_pup__Bool.lo] Error 1
make[2]: Leaving directory '/tmp/modules/yaksa'
make[1]: *** [Makefile:32250: all-recursive] Error 1
make[1]: Leaving directory '/tmp'
make: *** [Makefile:9251: all] Error 2
```

```
~/MPI/mpich/modules/yaksa$ git status
HEAD detached from c495759
```

```
~/MPI/mpich$ git log -n1
commit 209b07961111614abc4e9b06067c67855b20d3c7 (HEAD -> main)
```

**Comment by raffenet** [comment]
> I am seeing a problem with `compute_30`. I thought this was fixed already.

Can you send `config.log` from the yaksa directory? That particular commit hash should have the fix...

**Comment by jeffhammond** [comment]
I might have used a stale configure...

**Pull request merged**
